### PR TITLE
[Sprint: 41] XD-2574 - support multiple locators for gemfire modules

### PR DESCRIPTION
**Pull request by dturanski**
This allows the `host` and `port` properties to accept comma-delimited strings. The rule is that one or the other may be single valued or they must both contain lists of the same length. For example to specify multiple ports on the same host, you can do:

     --host=myhost.my.com  --port=10334,10335

Likewise you can do:

     --host=myhost1.my.com,myhost2.my.com --port=10334

And of course:

     --host=myhost1.my.com,myhost2.my.com --port=10334,10335

 which will apply corresponding host-port pairs in list order

### CODE REVIEW

**Comment by dturanski** [comment]
Note the same semantics also apply if to direct server connections if --useLocator is not set.

File: extensions/spring-xd-extension-gemfire/src/main/java/org/springframework/xd/gemfire/GemfireHostPortMixin.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+	private String port;
+
+	private boolean useLocator;
+
+	public String getHost() {
+		return host;
+	}
+
+	@ModuleOption("host name of the cache server or locator (if useLocator=true). May be a comma delimited list.")
```
Comment (cppwfs) [inline extensions/spring-xd-extension-gemfire/src/main/java/org/springframework/xd/gemfire/GemfireHostPortMixin.java]: Need to remove the period at the end of the description.  ModuleOptionsMetadataSanityTests tagged this on build.  Will fix on merge.

File: extensions/spring-xd-extension-gemfire/src/main/java/org/springframework/xd/gemfire/GemfireHostPortMixin.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	public void setHost(String host) {
+		this.host = host;
+	}
+
+	public String getPort() {
+		return port;
+	}
+
+	@ModuleOption("port of the cache server or locator (if useLocator=true). May be a comma delimited list.")
```
Comment (cppwfs) [inline extensions/spring-xd-extension-gemfire/src/main/java/org/springframework/xd/gemfire/GemfireHostPortMixin.java]: Need to remove the period at the end of the description. ModuleOptionsMetadataSanityTests tagged this on build. Will fix on merge.

File: extensions/spring-xd-extension-gemfire/src/main/java/org/springframework/xd/gemfire/GemfireCQSourceOptionsMetadata.java [hunk 3]
```diff
@@ -21,6 +21,7 @@
 
 import org.hibernate.validator.constraints.NotBlank;
 
+import org.springframework.xd.module.options.spi.Mixin;
```
Comment (cppwfs) [inline extensions/spring-xd-extension-gemfire/src/main/java/org/springframework/xd/gemfire/GemfireCQSourceOptionsMetadata.java]: Need to remove un-used imports.   Will cleanup on merge.

File: extensions/spring-xd-extension-gemfire/src/main/java/org/springframework/xd/gemfire/GemfireServerSinkModuleOptionsMetadata.java [hunk 4]
```diff
@@ -18,6 +18,9 @@
 
 import static org.springframework.xd.module.options.spi.ModulePlaceholders.XD_STREAM_NAME;
 
+import javax.validation.constraints.AssertTrue;
```
Comment (cppwfs) [inline extensions/spring-xd-extension-gemfire/src/main/java/org/springframework/xd/gemfire/GemfireServerSinkModuleOptionsMetadata.java]: Remove Unused imports  will fix on merge

File: extensions/spring-xd-extension-gemfire/src/main/java/org/springframework/xd/gemfire/GemfireSourceOptionsMetadata.java [hunk 5]
```diff
@@ -20,6 +20,7 @@
 
 import org.hibernate.validator.constraints.NotBlank;
 
+import org.springframework.xd.module.options.spi.Mixin;
```
Comment (cppwfs) [inline extensions/spring-xd-extension-gemfire/src/main/java/org/springframework/xd/gemfire/GemfireSourceOptionsMetadata.java]: Remove Unused imports will fix on merge

File: modules/common/gemfire-sink.groovy [hunk 6]
```diff
@@ -0,0 +1,10 @@
+beans {
```
Comment (cppwfs) [inline modules/common/gemfire-sink.groovy]: Just curious why we changed out the XML with Groovy?

File: modules/common/gemfire-sink.xml [hunk 7]
```diff
@@ -1,23 +0,0 @@
...
-		<gfe:pool id="client-pool">
-			<gfe:server host="${host}" port="${port}" />
-		</gfe:pool>
-	</beans>
-	<beans profile="use-locator">
-		<gfe:pool>
-			<gfe:locator host="${host}" port="${port}" />
-		</gfe:pool>
-	</beans>
-</beans>
```
Comment (cppwfs) [inline modules/common/gemfire-sink.xml]: end line nag... Fix on merge

File: modules/source/gemfire/config/gemfire.groovy [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+	gfe.'client-region'(id: 'region', 'cache-ref': 'client-cache', name: '${regionName}', 'data-policy': 'EMPTY') {
+		gfe.'key-interest' {
+			bean(id: 'key', 'class': 'java.lang.String') {
+				'constructor-arg'(value: 'ALL_KEYS')
+			}
+		}
+	}
+
+	si.channel(id: 'output')
```
Comment (cppwfs) [inline modules/source/gemfire/config/gemfire.groovy]: not sure we need to redeclare the output channel since it is already present in the gemfire-source

**Comment by cppwfs** [comment]
LGTM Starting Acceptance Test Pass.

**Comment by cppwfs** [comment]
LGTM Cleanup and Merge

**Pull request closed without merging**